src/AggregateFunctions/QuantileInterpolatedWeighted.h

```diff
@@ -0,0 +1,302 @@
+            {
+                idx = size - 2;
+            }
+            else
+            {
+                while (level > value_weight_pairs[idx + 1].second)
+                    idx++;
+            }
+        }
+
+        size_t l = idx;
+        size_t u = idx + 1 < size ? idx + 1 : idx;
+
+        Float64 xl = value_weight_pairs[l].second, xr = value_weight_pairs[u].second;
+        UnderlyingType yl = value_weight_pairs[l].first, yr = value_weight_pairs[u].first;
+
+        if (level < xl)
+            yr = yl;
+        if (level > xr)
+            yl = yr;
+
+        return static_cast<T>(interpolate(level, xl, xr, yl, yr));
+    }
+
+    /// Get the `size` values of `levels` quantiles. Write `size` results starting with `result` address.
+    /// indices - an array of index levels such that the corresponding elements will go in ascending order.
+    template <typename T>
+    void getManyImpl(const Float64 * levels, const size_t * indices, size_t num_levels, Value * result) const
+    {
+        size_t size = map.size();
+
+        if (0 == size)
+        {
+            for (size_t i = 0; i < num_levels; ++i)
+                result[i] = Value();
+            return;
+        }
+
+        using Pair = typename std::pair<UnderlyingType, Float64>;
+        std::vector<Pair> value_weight_pairs;
+
+        Float64 sum_weight = 0;
+        for (const auto & pair : map)
+        {
+            sum_weight += pair.getMapped();
+            auto value = pair.getKey();
+            auto weight = pair.getMapped();
+            value_weight_pairs.push_back(std::make_pair(value, weight));
+        }
+
+        ::sort(value_weight_pairs.begin(), value_weight_pairs.end(), [](const Pair & a, const Pair & b) { return a.first < b.first; });
+
+        Float64 accumulated = 0;
+
+        /// vector for populating and storing the cumulative sum using the provided weights.
+        /// example: [0,1,2,3,4,5] -> [0,1,3,6,10,15]
+        std::vector<Float64> weights_cum_sum;
+
+        bool first = true;
+        for (size_t idx = 0; idx < size; ++idx)
+        {
+            accumulated += value_weight_pairs[idx].second;
+
+            if (first)
+            {
+                weights_cum_sum.emplace_back(value_weight_pairs[idx].second);
+                first = false;
+            }
+            else
+            {
+                weights_cum_sum.emplace_back(accumulated);
+            }
+        }
+
+        /// The following estimation of quantile is general and the idea is:
+        /// https://en.wikipedia.org/wiki/Percentile#The_weighted_percentile_method
+
+        /// calculates a simple cumulative distribution based on weights
+        /// Note: if sum_weight is 0 then subsequent division is inf, so set the value of the exp to 0.
+        for (size_t idx = 0; idx < size; ++idx)
+            value_weight_pairs[idx].second
+                = sum_weight == 0 ? 0 : (weights_cum_sum[idx] - 0.5 * value_weight_pairs[idx].second) / sum_weight;
```

Comment:
Can weights be negative? Because if not then `sum_weight == 0` necessarily means all `value_weight_pairs[idx].second == 0`

Reply:
No, only unsigned integers are allowed.

Reply:
In this case, it only makes sense to go through elements of `value_weight_pairs` if `sum_weight != 0`
```suggestion
        if (sum_weight != 0)
        {
            for (size_t idx = 0; idx < size; ++idx)
                value_weight_pairs[idx].second = (weights_cum_sum[idx] - 0.5 * value_weight_pairs[idx].second) / sum_weight;
         }
```

Reply:
sure fixed.